src/core/parse_worker.hpp

```diff
@@ -0,0 +1,222 @@
+#pragma once
+
+#include <string>
+#include <unordered_set>
+#include <vector>
+#include <nlohmann/json.hpp>
+#include "symbol.hpp"
+#include "token.hpp"
+#include "parse_feeder.hpp"
+
+namespace docgen {
+namespace core {
+
+/*
+ * A ParseWorker object holds an ordered array of SymbolHandler objects;
+ * at any given point, only one of these SymbolHandlers is being handled.
+ *
+ * On every call to proc(), the current handler is checked for a match;
+ * so long as it isn't matched (waiting to match), proc() will call for
+ * the handler to handle_workers(), which will process based on
+ * its array of ParseWorker objects. When the current handler is matched,
+ * proc() will call for the handler to handle_token(), which executes
+ * the routine set by handler's passed function pointer; then, set to
+ * move on to the next handler. On the last handler, this will wrap around
+ * to the first, for so long as the passed "iters" limiter isn't
+ * exceeded (unlimited by default); once it is, this ParseWorker is said
+ * to be in a "done" state, on which proc() will no longer execute.
+ *
+ * Said to be in a "working" state when handling a SymbolHandler after
+ * the first; this only matters to a SymbolHandler which is holding
+ * this ParseWorker.
+ */
+class ParseWorker 
+{
+	public:
+		using worker_t = ParseWorker;
+		using worker_arr_t = std::vector<worker_t>;
+		using worker_arr_init_t = std::initializer_list<worker_t>;
+
+		using symbol_t = Symbol;
+		using token_t = Token<symbol_t>;
+
+		/*
+		 * A SymbolHandler object holds an unordered set of symbols to match against (symbols_),
+		 * a function pointer which references a routine to execute on match (on_symbol_), and
+		 * an array of ParseWorker objects to process while still unmatched (workers_).
+		 *
+		 * If neg_ is set true, a given symbol will only match if it is not in symbols_.
+		 *
+		 * If working_ is not a nullptr, it indicates that one of the workers is currently in
+		 * the "working" state, and points to this worker; until it is done, only this worker
+		 * will be processed.
+		 */
+		class SymbolHandler
+		{
+		public:	
+			using symbol_set_t = std::unordered_set<symbol_t>;
+			using symbol_arr_init_t = std::initializer_list<symbol_t>;
+
+			using routine_t = void (*)(const token_t&, ParseFeeder&);
```

Comment:
See my general comment first. This can then be changed to
```cpp
using routine_t = typename RoutineTraits::routine_t;
```

Reply:
I like the inheritance thing for something like `RoutineTraits`, so I will be implementing this—but I'll still have to do more research on the `std::function` thing. If it's really a one-line change, we could always benchmark this later on :)

Reply:
Yeah you don't have to do change it here but I at least want to know if it works so can tou at least just test that.